File: app/src/main/java/cjx/com/diary/view/activity/MainActivity.java
package cjx.com.diary.view.activity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;

import butterknife.BindView;
import butterknife.ButterKnife;
import cjx.com.diary.R;
import cjx.com.diary.base.BaseActivity;
import cjx.com.diary.presenter.impl.MainPresenterImp;
import cjx.com.diary.util.Utils;
import cjx.com.diary.view.fragment.FindFragment;
import cjx.com.diary.view.fragment.HomePageFragment;
import cjx.com.diary.view.fragment.PersonalFragment;
import cjx.com.diary.view.fragment.PhotoFragment;

public class MainActivity extends BaseActivity {
    @BindView(R.id.navigation)
    BottomNavigationView mNavigationView;

    FragmentManager fm;

    HomePageFragment homePageFragment;

    FindFragment findFragment;

    PhotoFragment photoFragment;

    PersonalFragment personalFragment;

    int mCurrentFragmentIndex = 0;

    public static void action(Context context) {
        context.startActivity(new Intent(context, MainActivity.class));
    }

    private MainPresenterImp mPresenter;

    private BottomNavigationView.OnNavigationItemSelectedListener mOnNavigationItemSelectedListener
            = item -> {
        switch (item.getItemId()) {
            case R.id.navigation_home:
                switchFragment(0);
                return true;
            case R.id.navigation_dashboard:
                switchFragment(1);
                return true;
            case R.id.navigation_photo:
                switchFragment(2);
                return true;
            case R.id.navigation_notifications:
                switchFragment(3);
                return true;
        }
        return false;
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        mPresenter = new MainPresenterImp();
        mPresenter.bindView(this, null);
        mNavigationView.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener);
        switchFragment(0);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        switchFragment(0);
    }

    /**
     * 切换fragment
     */
    public void switchFragment(int fragmentIndex) {
        mCurrentFragmentIndex = fragmentIndex;
        if (fm == null)
            fm = getSupportFragmentManager();
        FragmentTransaction transaction = fm.beginTransaction();
        if (homePageFragment == null) {
            homePageFragment = (HomePageFragment) HomePageFragment.newInstance();
            transaction.add(R.id.fl_container, homePageFragment);
        }
        if (findFragment == null) {
            findFragment = (FindFragment) FindFragment.newInstance();
            transaction.add(R.id.fl_container, findFragment);
        }
        if (photoFragment == null) {
            photoFragment = (PhotoFragment) PhotoFragment.newInstance();
            transaction.add(R.id.fl_container, photoFragment);
        }
        if (personalFragment == null) {
            personalFragment = (PersonalFragment) PersonalFragment.newInstance();
            transaction.add(R.id.fl_container, personalFragment);
        }
        transaction.hide(homePageFragment).hide(findFragment).hide(photoFragment).hide(personalFragment);
        switch (fragmentIndex) {
            case 0:
                transaction.show(homePageFragment);
                break;
            case 1:
                transaction.show(findFragment);
                break;
            case 2:
                transaction.show(photoFragment);
                break;
            case 3:
                transaction.show(personalFragment);
                break;
        }
        transaction.commitAllowingStateLoss();
        mNavigationView.getMenu().getItem(mCurrentFragmentIndex).setChecked(true);
    }


    private long mLastBackPress = 0;

    @Override
    public void onBackPressed() {
        long time = System.currentTimeMillis();
        if (time - mLastBackPress < 2000) {
            super.onBackPressed();
        } else {
            mLastBackPress = time;
            Utils.showToast(mActivity, getString(R.string.tip_exit));
        }
    }
}

File: app/src/main/java/cjx/com/diary/view/activity/MainActivity.kt
package cjx.com.diary.view.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.BottomNavigationView
import android.support.v4.app.FragmentManager
import butterknife.ButterKnife
import cjx.com.diary.R
import cjx.com.diary.base.BaseActivity
import cjx.com.diary.presenter.impl.MainPresenterImp
import cjx.com.diary.util.Utils
import cjx.com.diary.view.fragment.*
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : BaseActivity() {


    private var fm: FragmentManager? = null

     var homePageFragment: HomePageFragment? = null

     var findFragment: FindFragment? = null

     var weightManagerFragment: WeightManagerFragment? = null

     var photoFragment: PhotoFragment? = null

     var personalFragment: PersonalFragment? = null

     var mCurrentFragmentIndex = 0

    private var mPresenter: MainPresenterImp? = null

    private val mOnNavigationItemSelectedListener = BottomNavigationView.OnNavigationItemSelectedListener{ item ->
        when (item.itemId) {
            R.id.navigation_home -> {
                switchFragment(0)
                 true
            }
            R.id.navigation_dashboard -> {
                switchFragment(1)
                 true
            }
            R.id.navigation_weight -> {
                switchFragment(2)
                 true
            }
            R.id.navigation_photo -> {
                switchFragment(3)
                 true
            }
            R.id.navigation_notifications -> {
                switchFragment(4)
                 true
            }
        }
        false
    }


    private var mLastBackPress: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)
        mPresenter = MainPresenterImp()
        mPresenter?.bindView(this, null)
        navigation.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener)
        switchFragment(0)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        switchFragment(0)
    }

    /**
     * 切换fragment
     */
    private fun switchFragment(fragmentIndex: Int) {
        mCurrentFragmentIndex = fragmentIndex
        if (fm == null)
            fm = supportFragmentManager
        val transaction = fm!!.beginTransaction()
        if (homePageFragment == null) {
            homePageFragment = HomePageFragment.newInstance() as HomePageFragment
            transaction.add(R.id.fl_container, homePageFragment)
        }
        if (findFragment == null) {
            findFragment = FindFragment.newInstance() as FindFragment
            transaction.add(R.id.fl_container, findFragment)
        }
        if (weightManagerFragment == null) {
            weightManagerFragment = WeightManagerFragment()
            transaction.add(R.id.fl_container, weightManagerFragment)
        }
        if (photoFragment == null) {
            photoFragment = PhotoFragment.newInstance() as PhotoFragment
            transaction.add(R.id.fl_container, photoFragment)
        }
        if (personalFragment == null) {
            personalFragment = PersonalFragment.newInstance() as PersonalFragment
            transaction.add(R.id.fl_container, personalFragment)
        }
        transaction.hide(homePageFragment).hide(findFragment).hide(weightManagerFragment).hide(photoFragment).hide(personalFragment)
        when (fragmentIndex) {
            0 -> transaction.show(homePageFragment)
            1 -> transaction.show(findFragment)
            2 -> transaction.show(weightManagerFragment)
            3 -> transaction.show(photoFragment)
            4 -> transaction.show(personalFragment)
        }
        transaction.commitAllowingStateLoss()
        navigation.menu.getItem(mCurrentFragmentIndex).isChecked = true
    }

    override fun onBackPressed() {
        val time = System.currentTimeMillis()
        if (time - mLastBackPress < 2000) {
            super.onBackPressed()
        } else {
            mLastBackPress = time
            Utils.showToast(this, getString(R.string.tip_exit))
        }
    }

    companion object {

        fun action(context: Context) {
            context.startActivity(Intent(context, MainActivity::class.java))
        }
    }
}
